drivers/coredump/coredump_impl.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Meta Platforms, Inc. and its affiliates.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <debug/coredump.h>
+#include <drivers/coredump.h>
+
+#define DT_DRV_COMPAT zephyr_coredump
+
+struct coredump_config {
+	/* Length of memory_regions array */
+	int length;
+
+	/* Memory regions specified in device tree */
+	int memory_regions[];
+};
+
+struct coredump_data {
+	/* Memory regions registered at run time */
+	sys_slist_t region_list;
+};
+
+static void coredump_impl_dump(const struct device *dev)
+{
+	const struct coredump_config *config = dev->config;
+
+	/*
+	 * Add each memory region specified in device tree to the core dump,
+	 * the memory_regions array should contain two entries per region
+	 * containing the start address and size.
+	 */
+	if ((config->length > 0) && ((config->length % 2) == 0)) {
+		for (int i = 0; i < config->length; i += 2) {
+			uintptr_t start_address = config->memory_regions[i];
+			size_t size = config->memory_regions[i+1];
+
+#if defined(CONFIG_MMU)
+			mm_reg_t virt_addr;
+
+			device_map(&virt_addr, start_address, size, K_MEM_CACHE_NONE);
+			coredump_memory_dump(virt_addr, virt_addr + size);
```

Comment:
This is wrong. You do not want to map the MMIO region again using `device_map`, this is supposedly done by the driver at init time. You should use `DEVICE_MMIO_GET` instead.

Reply:
Yeah, I think I understand better now. Let me take Chris's suggestion and primarily use the dt bindings for "simple" memcpy operations at dump time and provide an optional callback so consumers can do more complex operations themselves (like DEVICE_MMIO_GET), that callback can be invoked at dump time